### PR TITLE
models train and inference services

### PR DESCRIPTION
**Pull request by aleksandrinvictor**
- Update models inference service. Add export to onnx.
- Add train service
- Add hydra configs

### CODE REVIEW

**Comment by Yorko** [comment]
Воспроизвелись обе модели на тачке hostkey, еще проверю потом на другой с GPU (hetzner). Если придираться, то CPU с tf-idf по сути перестал поддерживаться – такой огромный образ с cuda на CPU не соберется. Но во-первых, городить ифчики в docker-compose.yml вряд ли хочется, во-вторых, все же мы уже точно перешли с tf-idf на BERT, так что норм. Еще, конечно, в реальном проекте такой большой PR не смержат без тестов, но тут про тесты отдельно подумаем. 

Прогоню еще на тачке hetzner – если все заведется, одобрю. 

пс. по мелочи – добавил `static/model/prod/.gitkeep` чтоб прям по инструкции сработало.

**Comment by Yorko** [comment]
Воспроизвел tf-idf на машине hetzner без GPU. 

## Шаги следующие:

 - завел инстанс CX21 (2 vCPU, 4 GB RAM) c Ubuntu 20.04, `ssh root@<ID_ADDRESS>` (либо с ssh-ключом, либо временный пароль приходит)
 - установил [docker](https://docs.docker.com/engine/install/ubuntu/) с плагинами, v 20.10.16, build aa7e414
  - создал sudo-пользователя: `sudo useradd -s /bin/bash -d /home/<USER_NAME>/ -m -G sudo <USER_NAME>` 
  - добавл нового пльзователя в sudo и docker: `sudo usermod -aG sudo <USER_NAME> && sudo usermod -aG docker <USER_NAME>`
 - залогинился с новым юзером: `exit && ssh <USER_NAME>@<IP_ADDRESS>` 
 - прописал путь к docker-compose v2.5.0: `echo 'export PATH=/usr/libexec/docker/cli-plugins/:$PATH' >> ~/.bashrc && source ~/.bashrc`
 - `ssh-keygen`, ключ из ~/.ssh/id_rsa.pub добавил в [настройках GitHub](https://github.com/settings/keys)
 - `git clone git@github.com:crypto-sentiment/crypto_sentiment_demo_app.git && cd crypto_sentiment_demo_app`
 - `git checkout -b feature/models_train_inference && git pull origin feature/models_train_inference`
 - добавил конфиг подключения к БД: `echo '<LOOK_IT_UP_ON_NOTION>' > db_connection.ini`
 - перетащил данные с локальной машины: `scp <local_path_to_20190110_train_4500.csv> root@<IP_ADDRESS>:/home/<USERNAME>/crypto_sentiment_demo_app/data`
 - указал "tf_idf" (instead of "bert") в conf/config.yaml для defaults.model
 - (актуально только для машины без GPU) закоментил концовку `docker-compose.yml` про выделение GPU-ресурсов
 - (актуально только для машины без GPU) закоментил зависимости `nvidia::cudatoolkit=11.3.1`, ` pytorch::pytorch=1.10.0=py3.9_cuda11.3_cudnn8.2.0_0` и `pytorch::torchvision=0.11.1=py39_cu113` в файле `crypto_sentiment_demo_app/train/environment.yml` и вместо этого добавил `torch>=1.11.0` в секции pip-зависимостей
 - запустил обучение tf-idf-модели: `USER=$(id -u) GROUP=$(id -g) docker compose --profile train up --build` 
 - поднял все приложение: `docker compose --profile production up --build`

## Проблемки по ходу

Под рутом не завелось ("addgroup: The GID `0` is already in use." на шаге "RUN addgroup --gid $GROUP_ID user" в crypto_sentiment_demo_app/train/Dockerfile. Пробовал передать рандомные user & group IDs – Permission denied при попытке дампа модели в `static/models/dev`. Вылечилось тем, что создал нового sudo-юзера, добавил его в группу докер и запускал уже все из-под него.

**Review by Yorko — APPROVED** [review]
Крутая работа, bert нормально завелся на тачке hostkey, tf-idf обучил с нуля на новой машине hetzner – все шаги детально описал, включая сложности, возникшие по ходу (под рутом не завелось, какие-то cuda-специфичные зависимости выкидывал). 

Косяк, связанный не с этим PR, а с краулером: метрика "Today's average news sentiment" показывает 0%, т.к. краулер не сработал: ""https://bitcointicker.co/news/" недоступен (или борятся с таким краулингом)

**Comment by aleksandrinvictor** [comment]
> Воспроизвелись обе модели на тачке hostkey, еще проверю потом на другой с GPU (hetzner). Если придираться, то CPU с tf-idf по сути перестал поддерживаться – такой огромный образ с cuda на CPU не соберется. Но во-первых, городить ифчики в docker-compose.yml вряд ли хочется, во-вторых, все же мы уже точно перешли с tf-idf на BERT, так что норм. Еще, конечно, в реальном проекте такой большой PR не смержат без тестов, но тут про тесты отдельно подумаем.
> 
> Прогоню еще на тачке hetzner – если все заведется, одобрю.
> 
> пс. по мелочи – добавил `static/models/prod/.gitkeep` чтоб прям по инструкции сработало.

В c64077d разделил докер образы для обучения для cpu и gpu. Можно tf-idf на cpu обучить.

**Comment by Yorko** [comment]
А ONNX нормально работает с разреженными данными? Прошлый раз я его тыкал, tf-idf не поддерживался. Сейчас пишет, что word-level поддерживает, а char-level –  вряд ли `UserWarning: The conversion of CountVectorizer may not work. only tokenizer='word' is fully supported. You may raise an issue at https://github.com/onnx/sklearn-onnx/issues.`

**Comment by aleksandrinvictor** [comment]
> А ONNX нормально работает с разреженными данными? Прошлый раз я его тыкал, tf-idf не поддерживался. Сейчас пишет, что word-level поддерживает, а char-level – вряд ли `UserWarning: The conversion of CountVectorizer may not work. only tokenizer='word' is fully supported. You may raise an issue at https://github.com/onnx/sklearn-onnx/issues.`

Тоже читал, что может быть проблема с некоторыми моделями. Проверил tf-idf + логрег с текущим конфигом, mae на валидации между предсказаниями sklearn модели и onnx получилось ~0.00056. Accuracy val sklearn: 0.7145, onnx: 0.7156.

**Comment by Yorko** [comment]
@aleksandrinvictor можно мержить, спасибо за крутой вклад!
